app/javascript/src/PaymentGateways/braintree/BraintreeForm.jsx

```diff
@@ -0,0 +1,149 @@
+// @flow
+
+import React, { useState, useEffect, useRef } from 'react'
+import {
+  BraintreeUserFields,
+  BraintreeCardFields,
+  BraintreeBillingAddressFields,
+  BraintreeSubmitFields,
+  createHostedFieldsInstance,
+  create3DSecureInstance,
+  validationConstraints,
+  hostedFieldOptions,
+  veryfyCard
+} from 'PaymentGateways'
+import hostedFields from 'braintree-web/hosted-fields'
+import threeDSecure from 'braintree-web/three-d-secure'
+import { CSRFToken } from 'utilities/CSRFToken'
+import validate from 'validate.js'
+import './styles.css'
+
+import type { Node } from 'react'
+import type { BraintreeFormProps, BillingAddressData } from 'PaymentGateways'
+
+const BraintreeForm = ({
+  braintreeClient,
+  billingAddress,
+  threeDSecureEnabled,
+  formActionPath,
+  countriesList,
+  selectedCountryCode
+}: BraintreeFormProps): Node => {
+  const formRef = useRef<HTMLFormElement | null>(null)
+  // eslint-disable-next-line flowtype/no-weak-types
+  const [hostedFieldsInstance, setHostedFieldsInstance] = useState<any>(null)
+  const [braintreeNonceValue, setBraintreeNonceValue] = useState('')
+  const [billingAddressData, setBillingAddressData] = useState<BillingAddressData>(billingAddress)
+  const [isCardValid, setIsCardValid] = useState(false)
+  const [isFormValid, setIsFormValid] = useState(false)
+  const [formErrors, setFormErrors] = useState(validate(formRef, validationConstraints))
+  const [cardError, setCardError] = useState(null)
+  const [isAwaiting, setIsAwaiting] = useState(false)
+
+  useEffect(() => {
+    const getHostedFieldsInstance = async () => {
+      const HFInstance = await createHostedFieldsInstance(hostedFields, braintreeClient, hostedFieldOptions, setIsCardValid, setCardError)
+      setHostedFieldsInstance(HFInstance)
+    }
+    getHostedFieldsInstance()
+  }, [])
+
+  useEffect(() => {
+    if (braintreeNonceValue && formRef.current) {
+      formRef.current.submit()
+    }
+  }, [braintreeNonceValue])
+
+  useEffect(() => {
+    const formValid = isCardValid && !formErrors && !isAwaiting
+    setIsFormValid(formValid)
+  }, [formErrors, isCardValid, isAwaiting])
```

Comment:
`isCardValid`, `formErrors` and `isAwaiting` are all part of the state, therefore `isFormValid` is a pure value. By using `useEffect` the component will be rendered again only to calculate this.

There is no need for a `isFormValid` state either since it doesn't depend on external changes.

```suggestion
  const isFormValid = isCardValid && !formErrors && !isAwaiting
```